cmake/package_tooling.cmake

```diff
@@ -0,0 +1,128 @@
+# controlled with: '-B <path-to-build>'
+#
+# For example, if you were invoking CMake for 'hello_world' sample as:
+#   $ cmake -DBOARD=<board> -B build -S samples/hello_world
+#
+# you can invoke only dts module (and dependencies) as:
+#   $ cmake -DBOARD=<board> -B build -S samples/hello_world \
+#           -DMODULES=dts -P <ZEPHYR_BASE>/cmake/package_tooling.cmake
+#
+# It is also possible to pass additional build settings.
+# If you invoke CMake for 'hello_world' as:
+#
+#   $ cmake -DBOARD=<board> -B build -S samples/hello_world -DOVERLAY_CONFIG=foo.overlay
+#
+# you just add the same argument to tooling like:
+#   $ cmake -DBOARD=<board> -B build -S samples/hello_world -DOVERLAY_CONFIG=foo.overlay \
+#           -DMODULES=dts -P <ZEPHYR_BASE>/cmake/package_tooling.cmake
+#
+# Note: the samples CMakeLists.txt file is not processed by package tooling, so
+#       any 'set(<var> <value>)' specified before 'find_package(Zephyr)' must be
+#       manually applied, for example if the CMakeLists.txt contains:
+#          set(OVERLAY_CONFIG foo.overlay)
+#          find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
+#       the 'foo.overlay' must be specified using '-DOVERLAY_CONFIG=foo.overlay'
+
+cmake_minimum_required(VERSION 3.20)
+
+# add_custom_target and set_target_properties are not supported in script mode.
+# However, several Zephyr CMake modules create custom target for user convenience
+# like menuconfig, boards, shields, etc.
+# As we are not generating a build system with this tool, only running part of
+# the modules, then we simply override those functions to allow running those
+# modules.
+function(add_custom_target)
+  # This silence the error: 'add_custom_target command is not scriptable'
+endfunction()
+
+function(set_target_properties)
+  # This silence the error: 'set_target_properties command is not scriptable'
+endfunction()
+
+# Find last `-B` and `-S` instances.
+foreach(i RANGE ${CMAKE_ARGC})
+  if(CMAKE_ARGV${i} MATCHES "^-B(.*)")
+    set(argB ${CMAKE_MATCH_1})
+    set(argB_index ${i})
+  elseif()
+  elseif(CMAKE_ARGV${i} MATCHES "^-S(.*)")
+    set(argS_index ${i})
+    set(argS ${CMAKE_MATCH_1})
+  endif()
+endforeach()
+
+if(DEFINED argB_index)
+  if(DEFINED argB)
+    set(CMAKE_BINARY_DIR ${argB})
+  else()
+    # value of -B follows in next index
+    math(EXPR argB_value_index "${argB_index} + 1")
+    set(CMAKE_BINARY_DIR ${CMAKE_ARGV${argB_value_index}})
+  endif()
+endif()
+
+if(DEFINED argS_index)
+  if(DEFINED argS)
+    set(APPLICATION_SOURCE_DIR ${argS})
+  else()
+    # value of -S follows in next index
+    math(EXPR argS_value_index "${argS_index} + 1")
+    set(APPLICATION_SOURCE_DIR ${CMAKE_ARGV${argS_value_index}})
+  endif()
+endif()
+
+# List of base modules needed in order to ensure same behavior as the default
+# Zephyr build system during configure time.
+set(BASE_MODULES python west root zephyr_module boards shields arch configuration_files)
```

Comment:
Dependencies between our internal CMake scripts used to be encoded in one place. boilerplate.cmake.

Now it's encoded in three places. Each individual cmake script, zephyr_default.cmake, and package_tooling.cmake.

Can we get it down to two places for maintainability reasons?

I'm concerned that when a new module is added people will forget to add it to all three locations.

e.g. we could have package_tooling.cmake execute zephyr_default.cmake and have zephyr_default.cmake return when we had executed enough code. So in zephyr_default.cmake it would look like:

```
include(dts)
if(STOP_AFTER_DTS)
return()
endif()

include(kconfig)
if(STOP_AFTER_KCONFIG)
return()
endif()

```

Reply:
I see it's been re-organized.

But could we have both zephyr_default and package_tooling re-use this same list ?

Reply:
> Dependencies between our internal CMake scripts used to be encoded in one place. boilerplate.cmake.

and this corresponds to `zephyr_default.cmake` if you want to look at it from that perspective.

Now, you could argue that `zephyr_default.cmake` doesn't need to include Zephyr CMake modules if they will anyway be loaded by their dependencies, but that will make it harder to lookup at a single location to see what gets loaded.

So that's why I prefer to list all CMake modules in `zephyr_default.cmake`, so that it's visible when it gets loaded.

> Now it's encoded in three places. 
> Each individual cmake script,

Because those are modules intended for re-usability, so they might have additional strong dependencies.
This is what I have been trying to explain here: 
https://github.com/zephyrproject-rtos/zephyr/pull/41301#discussion_r777444109
https://github.com/zephyrproject-rtos/zephyr/pull/41301#discussion_r777553129

and as said in that PR, it's a first step in the process.

>  zephyr_default.cmake, and package_tooling.cmake.
> To reduce the number of places we list dependencies we could drop this list and instead have package_tooling.cmake execute zephyr_default.cmake and have zephyr_default.cmake return when we had executed enough code.

Now, this is where I agree it should be possible to improve, as to ensure `zephyr_default.cmake` and `package_tooling.cmake` is using a synced list.

Will take a look at this.

Reply:
Pushed a fix.

`find_package(Zephyr COMPONENTS <component>)` now supports indicating subparts to load.

This allows for constructs like this:
`find_package(Zephyr COMPONENTS zephyr_default:dts)`

which will load `zephyr_default` until the module `dts` is loaded.

This means `package_tooling` no longer need to have any knowledge on this.
It simply passes on the user arguments to find_package, and then let's zephyr_default do the magic.